tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.tla

```diff
@@ -0,0 +1,384 @@
+\* Copyright (c) 2024, Oracle and/or its affiliates.
+
+----------------------- MODULE BufferedRandomAccessFile -----------------------
+\* This is a model-checkable specification for BufferedRandomAccessFile.java.
+\* It covers the core fields as well as the seek, read, write, flush, and
+\* setLength operations.
+\*
+\* There are three major correctess conditions:
+\*
+\*   (1) the internal invariants V1-V5 should hold
+\*   (2) the behavior should refine a general RandomAccessFile
+\*   (3) each operation should refine its RandomAccessFile counterpart
+
+EXTENDS Naturals, Sequences, TLC, Common
+
+CONSTANT BuffSz
+
+VARIABLES
+    \* in-memory variables
+    dirty,
+    length,
+    curr,
+    lo,
+    buff,
+    diskPos,
+
+    \* the underlying file
+    file_content,
+    file_pointer
+
+vars == <<
+    dirty, length, curr, lo, buff, diskPos,
+    file_content, file_pointer>>
+
+TypeOK ==
+    /\ dirty \in BOOLEAN
+    /\ length \in Offset
+    /\ curr \in Offset
+    /\ lo \in Offset
+    /\ buff \in Array(Symbols \union {ArbitrarySymbol}, BuffSz)
+    /\ diskPos \in Offset
+
+    /\ file_content \in ArrayOfAnyLength(Symbols \union {ArbitrarySymbol})
+    /\ ArrayLen(file_content) <= MaxOffset
+    /\ file_pointer \in Offset
+
+-------------------------------------------------------------------------------
+\* Internal invariants (copied from comment in BufferedRandomAccessFile.java)
+
+RelevantBufferContent ==
+    ArraySlice(buff, 0, Min(BuffSz, length - lo))
+
+LogicalFileContent == \* denoted c(f) in .java file
+    IF ArrayLen(RelevantBufferContent) > 0
+    THEN WriteToFile(file_content, lo, RelevantBufferContent)
+    ELSE file_content
+
+DiskF(i) == \* denoted disk(f)[i] in .java file
+    IF i >= 0 /\ i < ArrayLen(file_content)
+    THEN ArrayGet(file_content, i)
+    ELSE ArbitrarySymbol
+
+BufferedIndexes == lo .. (Min(lo + BuffSz, length) - 1)
+
+Inv1 ==
+    /\ diskPos = file_pointer
+    /\ length = ArrayLen(LogicalFileContent)
+
+\* Inv2 is a bit special.  Most methods restore it just before they return.  It
+\* is generally restored by calling `seeek(curr)`.  But, that behavior is
+\* difficult to model in straight TLA+ because each method may modify variables
```

Comment:
I liked how Finn and co. did a mix of TLA+ and PlusCal in the [Inconsistency in Cosmos DB paper](https://arxiv.org/pdf/2210.13661.pdf) to play to the two's strengths and get clear multi-step behaviors.  It might be worth it to look at something like that for when variables need multiple changes, and then the invariants would only be expected to hold when the program counter is in safe state.

Reply:
I actually started with PlusCal, but even though it makes modeling easier it adds a lot of specification overhead. Specifically, verifying that the output of `read()` is correct might require an extra `read_output` variable. It gets a little ugly if `read()` is not a single action.

But, definitely something to keep in mind if this class becomes any more complex.